sharkfuser/include/fusilli/node/conv_node.h

```diff
@@ -203,6 +203,144 @@ class ConvFPropNode : public NodeCRTP<ConvFPropNode> {
   }
 };
 
+class ConvWGradNode : public NodeCRTP<ConvWGradNode> {
+public:
+  ConvWGradAttr convWGradAttr;
+
+  ConvWGradNode(ConvWGradAttr &&attr, const Context &ctx)
+      : NodeCRTP(ctx), convWGradAttr(std::move(attr)) {}
+
+  // MLIR assembly emitter helper methods.
+  std::string emitNodePreAsm() const override final;
+  std::string getOperandNamesAsm() const;
+  std::string getOperandTypesAsm() const;
+  std::string getResultNamesAsm() const;
+  std::string getResultTypesAsm() const;
+  std::string getStrideOpsAsm() const;
+  std::string getPaddingOpsAsm() const;
+  std::string getDilationOpsAsm() const;
+  std::string getPermuteDYOpsAsm() const;
+  std::string getPermuteXOpsAsm() const;
+  std::string getPermuteDWOpsAsm() const;
+  std::string getPermutedEmptyDWOpsAsm() const;
+
+  const std::string &getName() const override final {
+    return convWGradAttr.getName();
+  }
+  Type getType() const override final { return Type::WGrad; }
+
+  ErrorObject preValidateNode() const override final {
+    FUSILLI_LOG_LABEL_ENDL("INFO: Pre-Validating ConvWGradNode '"
+                           << convWGradAttr.getName() << "'");
+    FUSILLI_RETURN_ERROR_IF(convWGradAttr.getPadding().empty(),
+                            ErrorCode::AttributeNotSet, "Conv padding not set");
+    FUSILLI_RETURN_ERROR_IF(convWGradAttr.getStride().empty(),
+                            ErrorCode::AttributeNotSet, "Conv stride not set");
+    FUSILLI_RETURN_ERROR_IF(convWGradAttr.getDilation().empty(),
+                            ErrorCode::AttributeNotSet,
+                            "Conv dilation not set");
+
+    std::shared_ptr<TensorAttr> dyT = convWGradAttr.getDY();
+    std::shared_ptr<TensorAttr> xT = convWGradAttr.getX();
+    std::shared_ptr<TensorAttr> dwT = convWGradAttr.getDW();
+
+    // Ensure input and weight tensors are set.
+    FUSILLI_RETURN_ERROR_IF(!dyT, ErrorCode::AttributeNotSet,
+                            "Conv gradient tensor DY not set");
+    FUSILLI_RETURN_ERROR_IF(!xT, ErrorCode::AttributeNotSet,
+                            "Conv input tensor X not set");
+    FUSILLI_RETURN_ERROR_IF(!dwT, ErrorCode::AttributeNotSet,
+                            "Conv output tensor DW not set");
+
+    // Rank checks on DY and X tensors.
+    size_t dyRank = dyT->getDim().size();
+    size_t xRank = xT->getDim().size();
+    size_t wRank = dwT->getDim().size();
+
+    FUSILLI_RETURN_ERROR_IF(
+        dyRank < 3 || xRank < 3, ErrorCode::InvalidAttribute,
+        "Conv input tensors DY/X must have a rank of at least 3");
```

Comment:
This checks that they're rank >=3, but not that they match each other. Should we check that or is that not the case here?

Reply:
nvm again 🤦🏻